How to get a more friendly error message in cc?

I'm looking for a way to get more user-friendly error messages to help diagnose issues in my C/C++ code or the bridging code between Rust and C. How can I achieve this?

Here's a simple example of what I'm facing:

Suppose you have a C file intended for use in Rust. You would typically compile it in build.rs using the cc crate. If the C file is correct, everything works fine. However, in the example below, the error message is quite confusing.
![Image](https://github.com/user-attachments/assets/5897883a-b8d2-4d3e-8a70-aaee13a63879)
![Image](https://github.com/user-attachments/assets/7d2584e8-348a-41f6-b0f2-06d39664cdc7)

You can see that cc simply reports "error occurred" without providing further details. In contrast, compiling the same file with gcc highlights the specific error line.Is there a way to configure the cc crate to provide detailed gcc or msvc error messages when there are issues in my C/C++ code? How can I set this up?

> Hmm, does cl.exe print error to stderr or stdout?
> 
> For stdout we just set it to inherit the stdin of build-script, so it should work fine.
> 
> For stderr forward, cc simply just write `cargo:warning={msg}\n`
> 
> https://docs.rs/cc/latest/src/cc/command_helpers.rs.html#239
> 
> There's line splitting, but it also flushes the stderr at the end, and I think rust stdlib deal with the encoding stuff, so it should be fine...

In my experiment, cl.exe prints error to stdout.
``` Shell
# cl.exe
E:\\programme\\vs2019\\software\\VC\\Tools\\MSVC\\14.41.34120\\bin\\HostX64\\x64\\cl.exe src/test.c > stdout.log 2> stderr.log

## stdout.log
test.c
src/test.c(3): error C2143: 语法错误: 缺少“;”(在“}”的前面)

## stderr.log
E:\\programme\\vs2019\\software\\VC\\Tools\\MSVC\\14.41.34120\\bin\\HostX64\\x64\\cl.exe : 用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.41.34123 版
所在位置 行:1 字符: 1
+ E:\\programme\\vs2019\\software\\VC\\Tools\\MSVC\\14.41.34120\\bin\\H ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (用于 x64 的 Micros...器 19.41.34123 版:String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
版权所有(C) Microsoft Corporation。保留所有权利。
```
while gcc prints error to stderr
``` Shell
# gcc
gcc src/test.c > stdoutgcc.log 2> stderrgcc.log

## stdoutgcc.log          

## stderrgcc.log
gcc : src/test.c: In function 'main':
所在位置 行:1 字符: 1
+ gcc src/test.c > stdoutgcc.log 2> stderrgcc.log
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (src/test.c: In function 'main'::String) [], RemoteException
    + FullyQualifiedErrorId : NativeCommandError
 
src/test.c:2:13: error: expected ';' before '}' token
    2 |     return 0
      |             ^
      |             ;
    3 | }
      | ~  
```

> Hmm maybe we need to forward stdout as well as stderr for the `cl.exe`

After my research, I found that there are two main difficulties in supporting error capture and output for MSVC.

Firstly, UTF-8 encoding support on Windows is an experimental feature that is not enabled by default (this seems to have been the case since Windows 10 and has persisted for quite some time). As a result, many users in non-English speaking countries use PowerShell or CMD programs that are not UTF-8 encoded. The MSVC output in these terminals is also not in UTF-8 format. This makes handling stdout and stderr using Rust's `String` very difficult. Here is an example:

```rust
let mut s = String::new();
if let Err(e) = child.stdout.take().unwrap().read_to_string(&mut s) {
    eprintln!("Failed to read child stdout: {}", e);
} else {
    eprintln!("child stdout: {}", s);
}
```
This works well for UTF-8 encoded terminals. However, in non-UTF-8 encoded terminals, it enters the else branch and fails to correctly output MSVC errors.

The second difficulty is that the forward_available function in command_helpers.rs currently only supports capturing and outputting stderr. In the settings, stdout seems to be either forwarded or discarded in most cases (depending on the .cargo_output(bool) setting). To capture the output, these codes need to be modified.

Capturing MSVC's stdout and outputting it in a similar manner to stderr seems to be quite challenging, so I feel that directly forwarding the output is a simpler choice.

However, the problem then becomes, why does forwarding not work effectively on Windows? This is a very strange issue.
Here is my experiment:
``` shell
PS C:\Users\xs\Desktop\test_cc> cargo build
   Compiling test_cc v0.1.0 (C:\Users\xs\Desktop\test_cc)
error: failed to run custom build command for `test_cc v0.1.0 (C:\Users\xs\Desktop\test_cc)`                                                                                                                                                                                         

Caused by:
  process didn't exit successfully: `C:\Users\xs\Desktop\test_cc\target\debug\build\test_cc-fd694f9983ca928f\build-script-build` (exit code: 1)
  --- stderr


  error occurred: Command "E:\\programme\\vs2019\\software\\VC\\Tools\\MSVC\\14.41.34120\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "-W4" "-Fo./2e40c9e35e9506f4-test.o" "-c" "src/test.c" with args cl.exe did not execute successfully (status code exit code: 2).

PS C:\Users\xs\Desktop\test_cc> chcp
活动代码页: 936

//  change encoding to utf-8
PS C:\Users\xs\Desktop\test_cc> chcp 65001
Active code page: 65001

PS C:\Users\xs\Desktop\test_cc> cargo build
   Compiling test_cc v0.1.0 (C:\Users\xs\Desktop\test_cc)
error: failed to run custom build command for `test_cc v0.1.0 (C:\Users\xs\Desktop\test_cc)`

Caused by:
  process didn't exit successfully: `C:\Users\xs\Desktop\test_cc\target\debug\build\test_cc-fd694f9983ca928f\build-script-build` (exit code: 1)
  --- stdout
  OPT_LEVEL = Some(0)
  TARGET = Some(x86_64-pc-windows-msvc)
  cargo:rerun-if-env-changed=VCINSTALLDIR
  VCINSTALLDIR = None
  cargo:rerun-if-env-changed=VSTEL_MSBuildProjectFullPath
 .....
  cargo:rerun-if-env-changed=CFLAGS
  CFLAGS = None
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  **_test.c
  src/test.c(3): error C2143: syntax error: missing ';' before '}'_**

  --- stderr
  

  error occurred: Command "E:\\programme\\vs2019\\software\\VC\\Tools\\MSVC\\14.41.34120\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MD" "-Z7" "-Brepro" "-W4" "-Fo./2e40c9e35e9506f4-test.o" "-c" "src/test.c" with args cl.exe did not execute successfully (status code exit code: 2).
```

So, what happened?

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1785,8 +1785,6 @@ impl Build {
             self.fix_env_for_apple_os(&mut cmd)?;
         }
 
-        disable_localization(&mut cmd);
-
         Ok((cmd, name))
     }
 
@@ -1951,6 +1949,8 @@ impl Build {
             let warnings_to_errors_flag = cmd.family.warnings_to_errors_flag().into();
             cmd.push_cc_arg(warnings_to_errors_flag);
         }
+
+       disable_localization(&mut cmd);
 
         Ok(cmd)
     }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1784,9 +1784,8 @@ impl Build {
         if cfg!(target_os = "macos") {
             self.fix_env_for_apple_os(&mut cmd)?;
         }
-        if msvc {
-            disable_localization(&mut cmd);
-        }
+
+        disable_localization(&mut cmd);
 
         Ok((cmd, name))
     }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1958,7 +1958,7 @@ impl Build {
         // This should be acceptable because other messages from rustc are in English anyway,
         // and may also be desirable to improve searchability of the linker diagnostics.
         if matches!(cmd.family, ToolFamily::Msvc { clang_cl: false }) {
-            cmd.env.push(("VSLANG".into(), "1033".into())); 
+            cmd.env.push(("VSLANG".into(), "1033".into()));
         } else {
             cmd.env.push(("LC_ALL".into(), "C".into()));
         }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1957,14 +1957,10 @@ impl Build {
         // and prevent inspection of msvc output in case of errors, which we occasionally do.
         // This should be acceptable because other messages from rustc are in English anyway,
         // and may also be desirable to improve searchability of the linker diagnostics.
-       
-        // No harm in setting both env vars simultaneously.
-        // Unix-style linkers.
         if matches!(cmd.family, ToolFamily::Msvc { clang_cl: false }) {
-            cmd.env.push(("LC_ALL".into(), "C".into()));
+            cmd.env.push(("VSLANG".into(), "1033".into())); 
         } else {
-            // MSVC's `link.exe`.
-            cmd.env.push(("VSLANG".into(), "1033".into()));
+            cmd.env.push(("LC_ALL".into(), "C".into()));
         }
 
         Ok(cmd)

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1784,6 +1784,9 @@ impl Build {
         if cfg!(target_os = "macos") {
             self.fix_env_for_apple_os(&mut cmd)?;
         }
+        if msvc {
+            disable_localization(&mut cmd);
+        }
 
         Ok((cmd, name))
     }
@@ -4092,6 +4095,21 @@ fn check_disabled() -> Result<(), Error> {
         ));
     }
     Ok(())
+}
+
+/// Copied from https://github.com/rust-lang/rust/blob/5db81020006d2920fc9c62ffc0f4322f90bffa04/compiler/rustc_codegen_ssa/src/back/linker.rs#L27-L38
+///
+/// Disables non-English messages from localized linkers.
+/// Such messages may cause issues with text encoding on Windows
+/// and prevent inspection of msvc output in case of errors, which we occasionally do.
+/// This should be acceptable because other messages from rustc are in English anyway,
+/// and may also be desirable to improve searchability of the linker diagnostics.
+fn disable_localization(cmd: &mut Command) {
+    // No harm in setting both env vars simultaneously.
+    // Unix-style linkers.
+    cmd.env("LC_ALL", "C");
+    // MSVC's `link.exe`.
+    cmd.env("VSLANG", "1033");
 }
 
 #[cfg(test)]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -4097,7 +4097,7 @@ fn check_disabled() -> Result<(), Error> {
     Ok(())
 }
 
-/// Copied from https://github.com/rust-lang/rust/blob/5db81020006d2920fc9c62ffc0f4322f90bffa04/compiler/rustc_codegen_ssa/src/back/linker.rs#L27-L38
+/// Copied from <https://github.com/rust-lang/rust/blob/5db81020006d2920fc9c62ffc0f4322f90bffa04/compiler/rustc_codegen_ssa/src/back/linker.rs#L27-L38>
 ///
 /// Disables non-English messages from localized linkers.
 /// Such messages may cause issues with text encoding on Windows

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1950,7 +1950,22 @@ impl Build {
             cmd.push_cc_arg(warnings_to_errors_flag);
         }
 
-       disable_localization(&mut cmd);
+        // Copied from <https://github.com/rust-lang/rust/blob/5db81020006d2920fc9c62ffc0f4322f90bffa04/compiler/rustc_codegen_ssa/src/back/linker.rs#L27-L38>
+        //
+        // Disables non-English messages from localized linkers.
+        // Such messages may cause issues with text encoding on Windows
+        // and prevent inspection of msvc output in case of errors, which we occasionally do.
+        // This should be acceptable because other messages from rustc are in English anyway,
+        // and may also be desirable to improve searchability of the linker diagnostics.
+       
+        // No harm in setting both env vars simultaneously.
+        // Unix-style linkers.
+        if matches!(cmd.family, ToolFamily::Msvc { clang_cl: false }) {
+            cmd.env.push(("LC_ALL".into(), "C".into()));
+        } else {
+            // MSVC's `link.exe`.
+            cmd.env.push(("VSLANG".into(), "1033".into()));
+        }
 
         Ok(cmd)
     }
@@ -4094,21 +4109,6 @@ fn check_disabled() -> Result<(), Error> {
         ));
     }
     Ok(())
-}
-
-/// Copied from <https://github.com/rust-lang/rust/blob/5db81020006d2920fc9c62ffc0f4322f90bffa04/compiler/rustc_codegen_ssa/src/back/linker.rs#L27-L38>
-///
-/// Disables non-English messages from localized linkers.
-/// Such messages may cause issues with text encoding on Windows
-/// and prevent inspection of msvc output in case of errors, which we occasionally do.
-/// This should be acceptable because other messages from rustc are in English anyway,
-/// and may also be desirable to improve searchability of the linker diagnostics.
-fn disable_localization(cmd: &mut Command) {
-    // No harm in setting both env vars simultaneously.
-    // Unix-style linkers.
-    cmd.env("LC_ALL", "C");
-    // MSVC's `link.exe`.
-    cmd.env("VSLANG", "1033");
 }
 
 #[cfg(test)]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -1956,7 +1956,7 @@ impl Build {
         // Such messages may cause issues with text encoding on Windows
         // and prevent inspection of msvc output in case of errors, which we occasionally do.
         // This should be acceptable because other messages from rustc are in English anyway,
-        // and may also be desirable to improve searchability of the linker diagnostics.
+        // and may also be desirable to improve searchability of the compiler diagnostics.
         if matches!(cmd.family, ToolFamily::Msvc { clang_cl: false }) {
             cmd.env.push(("VSLANG".into(), "1033".into()));
         } else {

